rust-runtime/aws-smithy-http/src/body.rs

```diff
@@ -191,6 +196,22 @@ impl SdkBody {
             f(self)
         }
     }
+
+    /// Given a function to modify an `SdkBody`, run that function against this `SdkBody` before
+    /// returning the result. **This function MUST NOT alter the contents of the body.**
+    pub fn map_preserve_contents(
```

Comment:
I think this could use more documentation. I'm still not sure what this is for or why I should use it from reading this doc comment.

Also, is there any way we can assert the content didn't change in debug mode? Probably not?